Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -171,26 +164,32 @@ class ThreadSafeWeakHashSet final {
             callback(item.get());
     }
 
+    unsigned sizeIncludingEmptyEntriesForTesting()
+    {
+        Locker locker { m_lock };
+        return m_map.size();
+    }
+
 private:
     ALWAYS_INLINE void moveFrom(ThreadSafeWeakHashSet&& other)
     {
         Locker locker { m_lock };
         Locker otherLocker { other.m_lock };
-        m_set = std::exchange(other.m_set, { });
+        m_map = std::exchange(other.m_map, { });
         m_operationCountSinceLastCleanup = std::exchange(other.m_operationCountSinceLastCleanup, 0);
     }
 
     ALWAYS_INLINE void amortizedCleanupIfNeeded() const WTF_REQUIRES_LOCK(m_lock)
     {
-        if (++m_operationCountSinceLastCleanup / 2 > m_set.size()) {
-            m_set.removeIf([] (auto& pair) {
-                return pair.first->objectHasStartedDeletion();
+        if (++m_operationCountSinceLastCleanup / 2 > m_map.size() || m_operationCountSinceLastCleanup > 512) {
```

Comment:
This introduces O(n^2) algorithms where we used to have O(n) algorithms when adding lots of items.  It is no longer amortized.
What if instead we kept another integer m_amortizedCleanupThreshold that started at 512, and upon cleanup reset to m_map.size() if we found anything to clean up and doubled if we didn't?